### PR TITLE
Parse lifetimes that start with a number and give specific error

### LINKED ISSUE #58786: Identify incorrect lifetimes using a number as identifier in the parser
```
error: character literal may only contain one codepoint
 --> src/main.rs:1:10
  |
1 | struct S<'1> { s: &'1 usize }
  |          ^^^^^^^^^^^
help: if you meant to write a `str` literal, use double quotes
  |
1 | struct S<"1> { s: &"1 usize }
  |          ^^^^^^^^^^^

error: unexpected close delimiter: `}`
 --> src/main.rs:1:29
  |
1 | struct S<'1> { s: &'1 usize }
  |                             ^ unexpected close delimiter
```

[should instead be](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2beb7b028278cdbb901fa1a5796f7a57)

```
error: lifetimes must be valid identifiers, not a numberic
 --> src/main.rs:1:10
  |
1 | struct S<'1> { s: &'1 usize }
  |          ^^        ^^ numbers are not a valid lifetime identifier
  |          |
  |          numbers are not a valid lifetime identifier
help: if you meant to write a valid lifetime, use a valid identifier
  |
1 | struct S<'a> { s: &'a usize }
  |          ^^        ^^
```

### PR DESCRIPTION
**Pull request by estebank**
Fix #58786.

### CODE REVIEW

File: src/libsyntax/parse/lexer/mod.rs [hunk 1]
```diff
@@ -1873,13 +1885,15 @@ fn is_block_doc_comment(s: &str) -> bool {
     res
 }
 
+/// Determine wether `c` is a valid start for an ident. `include_numeric` is used for error
```
Comment (bjorn3) [inline src/libsyntax/parse/lexer/mod.rs]: There doesn't seem to be an `include_numeric`.

Comment (VirrageS) [reply src/libsyntax/parse/lexer/mod.rs]: nit: ‘wether’ -> ‘whether’

File: src/libsyntax/parse/lexer/mod.rs [hunk 2]
```diff
@@ -1873,13 +1885,14 @@ fn is_block_doc_comment(s: &str) -> bool {
...
 
+/// Determine whether `c` is a valid start for an ident.
 fn ident_start(c: Option<char>) -> bool {
     let c = match c {
         Some(c) => c,
         None => return false,
     };
 
-    (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_' || (c > '\x7f' && c.is_xid_start())
+    (c.is_alphabetic() || c == '_' || (c > '\x7f' && c.is_xid_start()))
```
Comment (petrochenkov) [inline src/libsyntax/parse/lexer/mod.rs]: The point of this code and the code in `ident_continue` below was optimization for the common case.
In addition to that `is_alphabetic` is different from `is_xid_start`, so this is also a change in behavior.
I'd remove these changes.

File: src/libsyntax/parse/lexer/mod.rs [hunk 3]
```diff
@@ -1444,6 +1446,16 @@ impl<'a> StringReader<'a> {
                         self.mk_ident(&format!("'{}", lifetime_name))
                     });
 
+                    if c2.is_numeric() {
+                        // this is a recovered lifetime written `'1`, error but accept it
+                        self.err_span_(
+                            start_with_quote,
+                            self.pos,
+                            "lifetimes can't start with a number",
```
Comment (petrochenkov) [inline src/libsyntax/parse/lexer/mod.rs]:
```suggestion
                            "lifetime names cannot start with a number",
```

Comment (petrochenkov) [reply src/libsyntax/parse/lexer/mod.rs]: Nit: My impression was that diagnostics usually use the "cannot" form (confirmed to some degree by grep).

**Comment by petrochenkov** [comment]
r=me with comments addressed

File: src/libsyntax/parse/lexer/mod.rs (hunk 2 above)
Comment (estebank) [reply src/libsyntax/parse/lexer/mod.rs]: I thought that both `is_alphabetic` and `is_xid_start` had those optimizations as well, but it seems like it's not the case. I'll change it.

File: src/libsyntax/parse/lexer/mod.rs [hunk 4]
```diff
@@ -1888,8 +1903,11 @@ fn ident_continue(c: Option<char>) -> bool {
         None => return false,
     };
 
-    (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_' ||
-    (c > '\x7f' && c.is_xid_continue())
+    ((c >= 'a' && c <= 'z') ||
+        (c >= 'A' && c <= 'Z') ||
+        c.is_numeric() ||
```
Comment (petrochenkov) [inline src/libsyntax/parse/lexer/mod.rs]: Could you revert this as well (keep the old code exactly).

**Pull request merged**